src/engine/Service.cpp

```diff
@@ -0,0 +1,175 @@
+  // of IDs from this endpoint and IDs from the remote endpoint (the latter with
+  // a bitmask, to that we can easily distinguish them).
+  parsedQuery::SparqlValues sparqlValues;
+  std::string variables;
+  if (!std::getline(tsvResult, variables)) {
+    throw std::runtime_error(absl::StrCat("Response from SPARQL endpoint ",
+                                          serviceUrl.host(), " is empty"));
+  }
+  LOG(INFO) << "First line of response body: " << variables << std::endl;
+  for (const std::string_view& variableName : absl::StrSplit(variables, "\t")) {
+    sparqlValues._variables.push_back(Variable{std::string{variableName}});
+  }
+  // TODO: This duplicates code in SparqlQleverVisitor.cpp . Maybe it would be
+  // more meaningful to postpone the interpretation of the values from a VALUES
+  // clause (via `parseTripleObject`) to parsedQuery::Values ?
+  std::string line;
+  std::string lastLine;
+  while (lastLine = std::move(line), std::getline(tsvResult, line)) {
+    if (sparqlValues._values.size() == 0) {
+      LOG(INFO) << "Second line of response body: " << line << std::endl;
+    }
+    std::vector<TripleComponent> valueTuple;
+    for (auto value : absl::StrSplit(line, "\t")) {
+      valueTuple.push_back(
+          TurtleStringParser<TokenizerCtre>::parseTripleObject(value));
+    }
+    sparqlValues._values.push_back(valueTuple);
+  }
+  if (sparqlValues._values.size() > 1) {
+    LOG(INFO) << "Last line of response body: " << lastLine << std::endl;
+  }
+  resultAsValues_ = std::make_unique<Values>(qec, std::move(sparqlValues));
+}
+
+// ____________________________________________________________________________
+string Service::asStringImpl(size_t indent) const {
+  std::ostringstream os;
+  for (size_t i = 0; i < indent; ++i) {
+    os << " ";
+  }
+  // TODO: This duplicates code in GraphPatternOperation.cpp .
+  os << "SERVICE " << parsedServiceClause_.serviceVarOrIri_ << " {\n"
+     << parsedServiceClause_.servicePrologue_ << "\n"
+     << parsedServiceClause_.serviceQueryBody_ << "\n}\n";
+  return std::move(os).str();
+}
+
+// ____________________________________________________________________________
+string Service::getDescriptor() const {
+  std::ostringstream os;
+  os << "Service with IRI " << parsedServiceClause_.serviceVarOrIri_;
+  // << " and body " << parsedServiceClause_.serviceQueryBody_;
+  return std::move(os).str();
+}
+
+// ____________________________________________________________________________
+size_t Service::getResultWidth() const {
+  return resultAsValues_->getResultWidth();
+}
+
+// ____________________________________________________________________________
+vector<size_t> Service::resultSortedOn() const {
+  return resultAsValues_->resultSortedOn();
+}
+
+// ____________________________________________________________________________
+VariableToColumnMap Service::computeVariableToColumnMap() const {
+  return resultAsValues_->computeVariableToColumnMap();
+}
+
```

Comment:
I understand why you chose the `compute in constructor and put into VALUES clause` approach....
especially the `getResultWidth()` and `computeVariableToColumnMap()` would require additional logic in the parser (but nothing too bad).
In general I want to know from you, whether you want to leave it as is currently for someone else to work on it in the future, 
or whether you are willing to put some more time and effort in yourself.

Reply:
The values from the TSV result are now written directly to an `IdTable`, without a detour via the `Values` class.